internal/backend/state_actions.go

```diff
@@ -2,39 +2,72 @@ package backend
 
 import (
 	"context"
+	"fmt"
 	"strings"
 	"time"
 
+	"github.com/ProtonMail/gluon/rfc822"
+
 	"github.com/ProtonMail/gluon/imap"
 	"github.com/ProtonMail/gluon/internal/backend/ent"
 	"github.com/bradenaw/juniper/xslices"
 	"golang.org/x/exp/slices"
 )
 
-func (state *State) actionCreateMailbox(ctx context.Context, tx *ent.Tx, name string) (*ent.Mailbox, error) {
-	res, err := state.remote.CreateMailbox(state.metadataID, strings.Split(name, state.delimiter))
+func (state *State) actionCreateAndGetMailbox(ctx context.Context, tx *ent.Tx, name string) (*ent.Mailbox, error) {
+	internalID, res, err := state.remote.CreateMailbox(ctx, state.metadataID, strings.Split(name, state.delimiter))
 	if err != nil {
 		return nil, err
 	}
 
-	if err := state.apply(ctx, tx, imap.NewMailboxCreated(res)); err != nil {
+	exists, err := DBMailboxExistsWithID(ctx, tx.Client(), internalID)
+	if err != nil {
 		return nil, err
 	}
 
-	return DBGetMailboxByID(ctx, tx.Client(), res.ID)
+	if !exists {
+		mbox, err := DBCreateMailbox(
+			ctx,
+			tx,
+			internalID,
+			res.ID,
+			strings.Join(res.Name, state.delimiter),
+			res.Flags,
+			res.PermanentFlags,
+			res.Attributes,
+		)
+
+		if err != nil {
+			return nil, err
+		}
+
+		return mbox, nil
+	}
+
+	return DBGetMailboxByID(ctx, tx.Client(), internalID)
+}
+
+func (state *State) actionCreateMailbox(ctx context.Context, tx *ent.Tx, name string) error {
+	internalID, res, err := state.remote.CreateMailbox(ctx, state.metadataID, strings.Split(name, state.delimiter))
+	if err != nil {
+		return err
+	}
+
+	return DBCreateMailboxIfNotExists(ctx, tx, internalID, res, state.delimiter)
```

Comment:
Can you use this function in `actionCreateAndGetMailbox` as well?

Reply:
In this case we need access to the created `ent.Mailbox` data. That function does not return any mailbox on purpose to avoid loading the data from disk if not required.

There is some code duplication, but they achieve different things.

Reply:
fair enough, i wonder if there's a way to improve this (e.g. by having this function return a bool indicating if the mailbox was actually created), but let's forget this for now